src/app/SOCestimator/SOCestimator.c

```diff
@@ -0,0 +1,134 @@
+#include <stdbool.h>
+
+#include "common_macros.h"
+#include "BatteryCharacteristics.h"
+#include "SOCestimator.h"
+
+#define CELL_CAPACITY_Ah 6.550
+#define BATTERY_CAPACITY_Ah (NUM_SERIES_CELLS * CELL_CAPACITY_Ah)
+
+int SOC_raw;
+int SOC_limit;
+
+/**
+ * Use the average series cell voltage and ambient temperature to determine a maximum SOC.
+ * Voltage thresholds determined (shittily) by eyeballing voltage/discharge % graph on the cell datasheet.
+ * avg_V [in] - average series cell voltage
+ * ambient_temp_C [in] - ambient temperature
+ */
+static int SOC_limit_from_voltage(float avg_V, float ambient_temp_C)
+{
+    // determine which temperature range the ambient temp falls into
+    // saturate to safe values
+    int temp_index = SAT((ambient_temp_C + .5) / 10, 0, NUM_TEMP_RANGES - 1);
+    
+    if (avg_V < SOC_25_LIMIT_FOR_TEMP_INDEX(temp_index))
+    {
+        return 25;
+    }
+    else if (avg_V < SOC_50_LIMIT_FOR_TEMP_INDEX(temp_index))
+    {
+        return 50;
+    }
+    else if (avg_V < SOC_75_LIMIT_FOR_TEMP_INDEX(temp_index))
+    {
+        return 75;
+    }
+    else
+    {
+        return 100;
+    }
+}
+
+/**
+ * Calculate the change in SOC (delta SOC) resulting from a current held over an elapsed time.
+ * current_A [in] - current held
+ * time_elapsed_ms [in] - time elapsed
+ */
+static inline float calculate_dSOC(float current_A, unsigned int time_elapsed_ms)
+{
+    return (current_A * time_elapsed_ms * 0.001) / BATTERY_CAPACITY_Ah;
```

Comment:
Why do we multiply by 0.001? Is it a unit conversion?

Reply:
to seconds, need to be in SI units